packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts

```diff
@@ -0,0 +1,231 @@
+import { v4 as uuid } from 'uuid'
+
+import { getDefaultSlippageDecimalPercentageForSwapper } from '../../../constants'
+import type {
+  CommonTradeQuoteInput,
+  GetTradeRateInput,
+  ProtocolFee,
+  SwapErrorRight,
+  SwapperDeps,
+  TradeQuote,
+  TradeRate,
+} from '../../../types'
+import { SwapperName, TradeQuoteError } from '../../../types'
+import { getRate, makeSwapErrorRight } from '../../../utils'
+import { getJupiterQuote, isSupportedChainId } from '../utils/helpers'
+
+const _getTradeQuote = async (
+  input: CommonTradeQuoteInput,
+  deps: SwapperDeps,
+): Promise<Result<TradeQuote[], SwapErrorRight>> => {
+  const {
+    sellAsset,
+    buyAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmount,
+    affiliateBps,
+    receiveAddress,
+    accountNumber,
+    slippageTolerancePercentageDecimal,
+    isManualReceiveAddress,
+  } = input
+
+  const { assetsById } = deps
+
+  const jupiterUrl = deps.config.REACT_APP_JUPITER_API_URL
+
+  if (isManualReceiveAddress && buyAsset.assetId === solAssetId) {
+    return Err(
+      makeSwapErrorRight({
+        message: `unsupported manual receive address for SOL`,
+        code: TradeQuoteError.UnsupportedTradePair,
+      }),
+    )
+  }
+
+  if (!isSupportedChainId(sellAsset.chainId)) {
+    return Err(
+      makeSwapErrorRight({
+        message: `unsupported chainId`,
+        code: TradeQuoteError.UnsupportedChain,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
+  }
+
+  if (!isSupportedChainId(buyAsset.chainId)) {
+    return Err(
+      makeSwapErrorRight({
+        message: `unsupported chainId`,
+        code: TradeQuoteError.UnsupportedChain,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
+  }
+
+  const maybeQuoteResponse = await getJupiterQuote({
+    apiUrl: jupiterUrl,
+    sourceAsset: sellAsset.assetId === solAssetId ? wrappedSolAssetId : sellAsset.assetId,
+    destinationAsset: buyAsset.assetId === solAssetId ? wrappedSolAssetId : buyAsset.assetId,
+    commissionBps: affiliateBps,
+    amount: sellAmount,
+    slippageBps: convertDecimalPercentageToBasisPoints(
+      slippageTolerancePercentageDecimal ??
+        getDefaultSlippageDecimalPercentageForSwapper(SwapperName.Jupiter),
+    ).toFixed(),
+  })
+
+  if (maybeQuoteResponse.isErr()) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Quote request failed',
+        code: TradeQuoteError.NoRouteFound,
+      }),
+    )
+  }
+
+  const { data: quoteResponse } = maybeQuoteResponse.unwrap()
+
+  const getFeeData = async () => {
+    const { chainNamespace } = fromAssetId(sellAsset.assetId)
+
+    switch (chainNamespace) {
+      case CHAIN_NAMESPACE.Solana: {
+        const sellAdapter = deps.assertGetSolanaChainAdapter(sellAsset.chainId)
+        const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+          // Simulates a self-send, since we don't know the to just yet at this stage
+          to: input.sendAddress!,
```

Comment:
Unfortunately we cannot calculate the compute units before getting the quote, the only fees estimation we can do is a regular send, which will change execution time when we have all the jupiter instructions including compute units

Updated to use the `to` from the input instead but it doesn't change so much tbh as it's a regular send

We might have a small fee difference at this point because of the compute units that we can't calculate there...

Reply:
> Unfortunately we cannot calculate the compute units before getting the quote
> Also at rate time receive address is not defined as a part of the split PR we merged

This is `getTradeQuote` so we should definitely be able to? I think the main issue here may be WRT reusing the exact same logic for trades and quotes for Jupiter, looks like quotes should be able to leverage the full Tx data including full instructions set?

Reply:
It might be doable, but I think we shouldn't call the quote endpoint at rate time so we have to split them

Let me try something tomorrow:
- Adding the `getJupiterSwapInstructions` execution at quote and rate time so we can get a more realistic fees estimation

Reply:
Sounds much saner
> we shouldn't call the quote endpoint at rate time so we have to split them

That's absolutely the goal 🙏🏽 basically, if the logic really is the same for both, it's fine to abuse a quote endpoint to get a rate (a quote is really just a rate with defined BIP44 params).
But if they happen to differ even slightly, their logic should be consolidated somewhere sane and reused, so that either rate or quote can have some additional/different functionality (typically things like getting *real* fees estimation at quote time with real calldata is a good contender for this different logic, since depending on the swapper we may not know the `to` address, or approvals may not be granted which may result in reverts etc)

Reply:
Closing as split has been tackled proper